Remove coordinator and support forking

wks opened the pull request:
This PR removes the coordinator (a.k.a. controller) thread, and allows temporarily terminating and restarting all GC worker threads in order to support forking.

Major changes include:

-   `GCController` is removed.  All synchronization mechanisms involving the controller are removed, too.  Important synchronization operations, such as opening buckets and declaring GC finished, are done by the last parked worker.  The work packet `EndOfGC` is removed, and its job is now done by the last parked worker.
-   The `WorkerMonitor`, which previously synchronizes between GC workers, now also synchronizes between mutators and GC workers.  This allows mutators to trigger GC by directly communicating with GC workers.
-   Introduced a new mechanism: "goals".  Mutators can now request "goals", and GC workers will work on one goal at a time.  Currently, a "goal" can be either GC or StopForFork.  Triggering GC is now implemented as requesting the GC goal.
-   Added a pair of new APIs, namely `MMTK::prepare_to_fork()` and `MMTK::after_fork()`.  VM bindings call them before and after forking to let the MMTK instance do proper preparation for forking.

Fixes: https://github.com/mmtk/mmtk-core/issues/1053
Fixes: https://github.com/mmtk/mmtk-core/issues/1054

wks commented:
binding-refs
OPENJDK_BINDING_REPO=wks/mmtk-julia
OPENJDK_BINDING_REF=feature/fork
RUBY_BINDING_REPO=wks/mmtk-ruby
RUBY_BINDING_REF=feature/fork

k-sareen commented on src/memory_manager.rs:
```diff
@@ -510,6 +501,20 @@ pub fn initialize_collection<VM: VMBinding>(mmtk: &'static MMTK<VM>, tls: VMThre
+///
+/// Arguments:
+/// * `mmtk`: A reference to an MMTk instance.
+pub fn uninitialize_collection<VM: VMBinding>(mmtk: &'static MMTK<VM>) {
```
I can't say I like the name `uninitialize_collection` for this API. Can I suggest something like `prepare_to_fork` or just `pre_fork`? If you don't like mentioning `fork` in the API name, then maybe just `stop_gc_threads`?

wks replied:
After a second thought, I think this is a good idea.  The `initialize_collection` should be a once-per-MMTk-instance initializer.  Back in JikesRVM, `initialize_collection` was named `enableCollection` (JikesRVM doesn't allow enabling/disabling collection at run time).  It determines the number of threads, and initializes some space-specific data that depends on the number of threads (such as `spillMarkHistograms` in `Defrag`, which we now use `Mutex<Vec<Histogram>>` to avoid making it dependent on the number of GC workers).  It also initialize worker groups, for example, one worker group for parallel workers, and another worker group for concurrent workers, and it is similar to our lxr branch.

Given that `initialize_collection` should be called only once, it is worth making the distinction between `initialize_collection` and the `pre_fork` and `post_fork` functions that may happen as many times as the VM forks.

k-sareen replied:
Yes exactly. That's why I felt like `uninitialize_collection` was not reflective of the actual use-case of this API. I pretty much only envision this being used for `fork`, so I think `pre_fork` is appropriate, but if you think it might be used for other reasons, then perhaps just a generic name such as `stop_gc_threads` would be more appropriate.

wks commented:
> @wks Did you make a PR accidentally for Angus' fork of Ruby instead of MMTk's fork of Ruby?

Oops. Yes. That was embarrassing.

wks commented:
Timeout happened for the OpenJDK binding when running debug build.  I'll try to run it again and see if it is just randomly slow.

The JikesRVM binding crashed, but that error is reproducible using the master branch, so it is likely not related to this PR.  See https://github.com/mmtk/mmtk-jikesrvm/issues/108

k-sareen commented on src/memory_manager.rs:
```diff
@@ -486,15 +474,18 @@ pub fn start_control_collector<VM: VMBinding>(
-/// to create GC threads.
+/// is ready.  This call will invoke Collection::spawn_gc_thread() to create GC threads.
+///
+/// This function can also be called after `uninitialize_collection` is called in order to re-spawn
+/// the GC thtreads.
```
```suggestion
/// This function can also be called after `uninitialize_collection` in order to re-spawn
/// GC threads.
```

k-sareen commented on src/memory_manager.rs:
```diff
@@ -510,6 +501,20 @@ pub fn initialize_collection<VM: VMBinding>(mmtk: &'static MMTK<VM>, tls: VMThre
     probe!(mmtk, collection_initialized);
 }
 
+/// Ask GC worker threads to exit.  Currently, this function is for the sole purpose of forking.
```
I think it would be prudent to both mention in the API docs and test (via the MockVM) what happens in cases when:

1. A GC is required, but all GC threads have exited
2. A GC is in progress, and `pre_fork` is called

wks replied:
I changed the comments to make it clear that 

1.  The VM must not attempt to allocate between `prepare_to_fork` and `after_fork`.
2.  Doing GC and preparing for forking are mutually exclusive, and GC workers must finish one of them before starting the other.

wks commented:
The current code has a bug.  The function `WorkerGroup::wait_until_worker_exited` counts the number of `GCWorker` instances delivered, but at that time, the GC worker threads themselves may not have exited.  Such threads may still have cleaning-up to do, and other operations to execute.  For example, if a GC worker thread prints log after giving back the `GCWorker` struct, the main thread may think the worker has exited and calls `fork()`.  That will leave the lock on `Stderr` contended, and hang the process.  It is necessary to `join` the GC worker threads to ensure the OS-level threads actually exited.

qinsoon commented on src/mmtk.rs:
```diff
@@ -220,6 +220,56 @@ impl<VM: VMBinding> MMTK<VM> {
+    /// using the saved contexts.
+    ///
+    /// This function returns when all MMTK threads stopped.
+    pub fn prepare_to_fork(&'static self) {
```
`fork` is unix specific. We should either avoid using the term, or make it specific to certain targets. As we are not doing anything specific to forking in MMTk (we just stop GC threads), I think it may be a good idea to avoid using the term `fork`.

wks replied:
Yes.  `fork` is unix-specific.  But I think we should be very explicit about forking because it has very unusual requirements.  For example, when calling `fork()`, the process must have only one thread.  If the process has multiple threads, the only thing safe to do is calling `exec()` or other async-signal-safe functions (see https://man7.org/linux/man-pages/man2/fork.2.html).  The implementation of `prepare_to_fork` is centered around those requirements.  The `GCThreadJoinHandle` trait which I just introduced yesterday is one example.   It waits until the underlying native thread of a GC thread quits.  It may be very different if we simply want to, for example, grow or shrink the number of GC threads at run time.  In that case, it is sufficient to let the GC thread give up their context, but the VM may repurpose those threads as other threads (such as used as mutators).

And the name should be exposed through the API.  If a VM needs forking (Android ART and CRuby), it knows precisely what it is doing.  CRuby carefully brings down other helper threads, too, before forking.

I am OK with making it platform-specific.  VMs treat forking as platform-specific, too.  CRuby only supports `Kernel#fork` if the underlying OS supports `fork()`, and Andorid only runs on its own modified Linux kernel.  But I don't know how to do it well now.  We need some kind of platform abstraction layer.  Then we can label platforms that support `fork()`.  That includes Linux, Mac, some UNIX variants, and CYGWIN, too.

qinsoon replied:
> Yes. `fork` is unix-specific. But I think we should be very explicit about forking because it has very unusual requirements. For example, when calling `fork()`, the process must have only one thread. If the process has multiple threads, the only thing safe to do is calling `exec()` or other async-signal-safe functions (see https://man7.org/linux/man-pages/man2/fork.2.html). The implementation of `prepare_to_fork` is centered around those requirements. The `GCThreadJoinHandle` trait which I just introduced yesterday is one example. It waits until the underlying native thread of a GC thread quits. It may be very different if we simply want to, for example, grow or shrink the number of GC threads at run time. In that case, it is sufficient to let the GC thread give up their context, but the VM may repurpose those threads as other threads (such as used as mutators).

In this API function, does MMTk core do anything specific to forking other than just stopping GC threads? Is it the binding rather than MMTk core that needs to know the semantics of forking?

wks replied:
1.  It keeps the `GCWorker` struct instances so that GC workers can be re-spawn later, reusing those structs.  If we are adjusting the number of GC workers, we probably would drop those structs.  This probably doesn't matter because the GC workers are almost stateless when not running any work packets.
2.  It needs to ensure the underlying OS thread of the GC workers exit, not just returning from the function `memory_manager::start_worker`.  Of course the binding needs to cooperate by implementing the `GCThreadJoinHandle`.

There is an alternative design.  MMTk core provides a function to ask all GC threads to return from their entry-point functions, such as `memory_manager::start_worker`.  Then the VM bindings implement their own mechanisms to wait until their underlying OS threads exited.  By doing this, bindings no longer need to implement `GCThreadJoinHandle` and give it to mmtk-core.

qinsoon replied:
> 1. It keeps the `GCWorker` struct instances so that GC workers can be re-spawn later, reusing those structs.  If we are adjusting the number of GC workers, we probably would drop those structs.  This probably doesn't matter because the GC workers are almost stateless when not running any work packets.
> 2. It needs to ensure the underlying OS thread of the GC workers exit, not just returning from the function `memory_manager::start_worker`.  Of course the binding needs to cooperate by implementing the `GCThreadJoinHandle`.

It sounds to me that both are design choices rather than some requirements of `fork`.

wks replied:
It is OK to destruct `GCWorker` when forking.  We will need to construct them again later.  It is fair to say this is a design choice.  It is an appropriate choice for forking.

It is OK to move the responsibility of waiting for the OS threads to the binding.  But either mmtk-core or the binding must wait for OS threads to exit.  This is a hard requirement for `fork`.

wks commented:
I tried an alternative approach.  In https://github.com/mmtk/mmtk-ruby/pull/54, the mmtk-ruby binding records the `JoinHandle` instances of GC threads, and joins them after calling `MMTK::prepare_to_fork()`.  It works with a prior commit in this PR: https://github.com/mmtk/mmtk-core/pull/1067/commits/176ad475b0ba05d6c1024b2c345ed203dcd02448

The advantage are:
-   mmtk-core has less obligations, and
-   Other bindings don't need to be aware of GCThreadJoinHandle, and
-   If it is hard for a binding to provide an abstract `GCThreadJoinHandle` to mmtk-core, it can wait for native threads using its own thread-management mechanisms.

But at the same time, the API `MMTK::prepare_to_fork()` alone is no longer sufficient to ensure it is safe to call `fork()` after it returns.  The binding needs to join the native threads by itself.  So we need to be extra cautious about its documentation.

It is OK to rename it to something like `stop_gc_threads`, but that makes it less clear that it is for forking.  If we refactor MMTk to change the way we stop GC threads, we need to keep in mind that there are at least two bindings (CRuby and Android) that relies on forking, and we need one working mechanism for forking.  The name `prepare_to_fork`, on the other hand, makes it very clear that it needs to support forking.

wks commented:
I marked this PR and its associated bindings as ready for review.

- mmtk-core (this one): https://github.com/mmtk/mmtk-core/pull/1067
- mmtk-openjdk: https://github.com/mmtk/mmtk-openjdk/pull/268
- mmtk-ruby: https://github.com/mmtk/mmtk-ruby/pull/53
- mmtk-julia: https://github.com/mmtk/mmtk-julia/pull/127
- mmtk-jikesrvm: https://github.com/mmtk/mmtk-jikesrvm/pull/159
- mmtk-v8: https://github.com/mmtk/mmtk-v8/pull/82

wks commented on src/mmtk.rs:
```diff
@@ -220,6 +220,90 @@ impl<VM: VMBinding> MMTK<VM> {
+    /// Initialize the GC worker threads that are required for doing garbage collections.
+    /// This is a mandatory call for a VM during its boot process once its thread system
+    /// is ready.  This function should be called exactly once, and should be called before
+    /// allocating objects in the MMTk heap.
```
That makes sense, but I think the current API is a bit wound up w.r.t. `GlobalState::initialized` and `VM::VMCollection::is_collection_enabled()`.  If I read the code correctly, the current code does assume `GlobalState::initialized` to be true when allocating (if it is possible to trigger GC).  But we may do another refactoring to tidy up the interacction between `initialized` and `is_collection_enabled`.

`Space::acquire()` contains the following code:

```rs
        let should_poll = VM::VMActivePlan::is_mutator(tls) && VM::VMCollection::is_collection_enabled();
        let allow_gc = should_poll && self.common().global_state.is_initialized();

        if should_poll && self.get_gc_trigger().poll(false, Some(self.as_space())) {
            assert!(allow_gc, "GC is not allowed here: collection is not initialized (did you call initialize_collection()?).");
```

If `should_poll` (determined by `is_collection_enabled()`) is true, and `poll()` indicates that collection is required, it will assert `allow_gc` is true.  That will fail if `initialized` is false.  If we allow allocation before calling `initialize_collection()`, we will also require `is_collection_enabled()` to always return false before calling `initialize_collection()`, too.  That requirement has not yet been reflected by our API docs, but I think it complicates things anyway.

`memory_manager::gc_poll()` also asserts `is_initialized()` after `VM::VMCollection::is_collection_enabled()` returns `true`.  It is arguable that `gc_poll` is for triggering GC, not allocation, and that we can require the VM not to call `gc_poll()` before calling `initialize_collection()`, but it is not that obvious from the API.

There are also some places involving stress GC that call `is_initialized()` (but strangely they do not `is_collection_enabled()`).  Examples include `Allocator::alloc_slow_inline` and `GCTrigger::should_do_stress_gc_inner`.

So I think for the current code, it is the safest to require `initialize_collection()` before allocation.  But I'll raise another issue for that.

qinsoon replied:
I don't think we want to require `initialize_collection()` before allocation. If our code requires that, it sounds like a bug to me. The current behavior is that the bindings can allocate before `initialize_collection()`, but when the heap is full and the binding has not yet called `initialize_collection()`, it will fail. What we may want is probably that a binding can allocate as much as they want without triggering a GC before calling `initialize_collection()`, which is the same as Java MMTk.

wks replied:
OK.  For now, I can remove "and should be called before allocating objects in the MMTk heap" from the comment.

qinsoon replied:
I just encountered the assertion failure in Julia. I created an issue here https://github.com/mmtk/mmtk-core/issues/1096.

wks commented on src/mmtk.rs:
```diff
@@ -349,6 +433,8 @@ impl<VM: VMBinding> MMTK<VM> {
         self.state
             .internal_triggered_collection
             .store(true, Ordering::Relaxed);
+        // TODO: The current `GCRequester::request()` is probably incorrect for internally triggered GC.
```
With this PR applied, `GCRequester` is now used for *mutator* threads to communicate with *GC worker* threads (there is no coordinator).  It does so by setting a variable ("goal") in `WorkerMonitor` and notifying one GC worker.  For optimization, `GCRequester` also uses an `AtomicBool` to filter out consecutive requests from mutators.

But "internally triggered GC" is used to upgrade a concurrent GC to a STW GC.  That is not done by communication between mutators and GC threads, but by GC threads coordinating themselves.  They still communicate via shared data structures such as `WorkerMonitor`, but setting a "goal" is probably not the right way to perform the "upgrading" because GC threads only respond to one goal at a time.

I am still not sure how to do such "upgrading" in Rust MMTk.  The `lxr` branch is not using the `trigger_internal_collection_request` function, either, although `GCRequester::request(&self, concurrent: bool)` has an additional argument `concurrent: bool` in the `lxr` branch.

wks commented on src/scheduler/scheduler.rs:
```diff
@@ -80,36 +85,61 @@ impl<VM: VMBinding> GCWorkScheduler<VM> {
+
+    /// Surrender the `GCWorker` struct of a GC worker when it exits.
+    pub fn surrender_gc_worker(&self, worker: Box<GCWorker<VM>>) {
+        let all_surrendered = self.worker_group.return_gc_worker_struct(worker);
```
Obviously, that saves some allocation and initialization.  The native code may also save pointers to `GCWorker` instances as `VMWorkerThread`.  It is the safest and the most convenient to reuse the existing structures, although VMs still need to take care of setting thread-local variables (for PThread or C/C++'s `thread_local` variables).

Another difficulties is that we currently assume the number of GC threads is fixed.  We still have a fixed number of `WorkerShared` and `deque::Worker` instances created when the `WorkerGroup` is created.  We still need to reuse those data structures anyway, even if we destroy and recreate `GCWorker` instances.  That would be more complicated.

But one thing we can explore is the ability to dynamically scale the number of GC workers at run time.  In that way, we no longer assume a fixed number of workers, and we don't need to pre-create `WorkerShared` and related instances before spawning GC threads.  We can then simply scale the number of GC threads down to 0 before forking, and re-scale it up (maybe to a different number depending on how many processes the VM wishes to fork into) after forking.  That'll be another refactoring.

qinsoon replied:
> Obviously, that saves some allocation and initialization. 

I don't think allocation and initialization really matters here. Forking is heavy weight. Compared to that, creating a few structs for GC workers should be trivial.

> The native code may also save pointers to `GCWorker` instances as `VMWorkerThread`. It is the safest and the most convenient to reuse the existing structures, although VMs still need to take care of setting thread-local variables (for PThread or C/C++'s `thread_local` variables).

I don't think this is a valid reason either. After forking, all the GC threads exit. It is very unlikely that the binding somehow holds a reference to the GC threads and the GC workers.

> Another difficulties is that we currently assume the number of GC threads is fixed. We still have a fixed number of `WorkerShared` and `deque::Worker` instances created when the `WorkerGroup` is created. We still need to reuse those data structures anyway, even if we destroy and recreate `GCWorker` instances. That would be more complicated.

This is a valid point, based on the current code base. We create `WorkerGroup` before spawning GC workers. The reason is that plans need a reference to the scheduler, and the scheduler includes a worker group so we create the worker group before we create the plan. However, this is not necessary. We use `None` for the worker group when we create the scheduler, and only create the worker group when we spawn threads. In this way, we have to bear the cost of `unwrap` everytime we use worker group (which should be fine -- when a thread accesses worker group to modify other workers shared state, it is not expected to be cheap anyway). See https://github.com/qinsoon/mmtk-core/commit/1f495764f2502aba1b3c827fe9adb07960e7c244 for a prototype for this refactoring.

My concern is that there seems no valid point why we need the complexity to allow reusing the GC worker structs, and once we accept those changes, it may need extra efforts to remove it. I can submit a pul request to remove the pre-created `deque::Worker` list if you think that is reasonable.

wks replied:
> I don't think allocation and initialization really matters here. Forking is heavy weight. Compared to that, creating a few structs for GC workers should be trivial.

That makes sense.  If that matters, it won't be feasible to scale the number of GC workers at run time, either.

> I don't think this is a valid reason either. After forking, all the GC threads exit. It is very unlikely that the binding somehow holds a reference to the GC threads and the GC workers.

Yes.  It should be unlikely that the binding keep references to one `GCWorker` from many different places.

Currently, the Ruby binding holds a pointer to the `GCWorker` instance in a thread-local variable for each GC worker thread.  Mutator threads are not aware of `GCWorker` instances.  The Ruby binding also creates GC workers as Rust threads (i.e. creating threads using the Rust standard library).  So the implementation is simple for Ruby.  The binding keeps track of how many GC worker threads actually exited (using `JoinHandle`), and the GC worker threads themselves can just exit.

In the OpenJDK binding, `MMTkCollectorThread` inherits `NamedThread`, and constructing that will need to register it to some global lists.  And `MMTkCollectorThread` holds a pointer to the `GCWorker` instance (in `MMTkCollectorThread::third_party_heap_collector`.  It is arguable that GC workers are reachable from the global list.  But I don't think that matters.  When a GC worker exits, the `MMTkCollectorThread` always needs to be removed from the global list and destructed, even if we keep the `GCWorker` instance.  OpenJDK doesn't need support for `fork()`, but I guess Android will be similar.  Android ART must have already taken care of this to support Zygote.

So I think you are right.  Well-written VMs and bindings shouldn't have pointers to `GCWorker` instances all over the place.

> This is a valid point, based on the current code base. We create `WorkerGroup` before spawning GC workers. The reason is that plans need a reference to the scheduler, and the scheduler includes a worker group so we create the worker group before we create the plan. However, this is not necessary. We use `None` for the worker group when we create the scheduler, and only create the worker group when we spawn threads. In this way, we have to bear the cost of `unwrap` everytime we use worker group (which should be fine -- when a thread accesses worker group to modify other workers shared state, it is not expected to be cheap anyway). See [qinsoon@1f49576](https://github.com/qinsoon/mmtk-core/commit/1f495764f2502aba1b3c827fe9adb07960e7c244) for a prototype for this refactoring.
> 
> My concern is that there seems no valid point why we need the complexity to allow reusing the GC worker structs, and once we accept those changes, it may need extra efforts to remove it. I can submit a pul request to remove the pre-created `deque::Worker` list if you think that is reasonable.

I think it is a good idea to make a dedicated pull request to refactor the creation of GC workers and remove the pre-created `deque::Worker`.  I looked at your refactoring.  I think the extra level of `unwrap()` is OK, but we should use safe code instead of relying on `Arc::to_raw` (casting `&` to `&mut` has undefined behavior anyway).

There are details that can be discussed.  For example, whether we use `Mutex<Option<WorkerGroup>>` or let a `WorkerGroup` have a `Mutex<Vec<Arc<GCWorkerShared>>>` instead.  The latter allows us to be aggressive and make the number of GC workers adjustable at run time.  But the former is also reasonable if we have multiple worker groups (e.g. one for concurrent GC and another for STW GC) so that a worker group may not always be available.

We can also move the `WorkerGroup` to a shared synchronized data structure, such as `WorkerMonitor` which I extensively used in this PR for synchronization.  One feature of this PR is that when the last GC worker parked, that worker will have already gained exclusive access to the `WorkerMonitor` struct.

My concern is that in the current `master` branch, there are a few complex states such as `enum WorkerGroupState { Sleeping, Working }` which I created for synchronizing between the workers and the coordinator.  It may be easier to do the refactoring after removing the coordinator.

wks commented on src/scheduler/worker.rs:
```diff
@@ -394,16 +233,50 @@ impl<VM: VMBinding> GCWorker<VM> {
 }
 
-/// A worker group to manage all the GC workers (except the coordinator worker).
+enum WorkerCreationState<VM: VMBinding> {
```
Yes.  Since the enum variant names already indicate the current state, it is much clearer to simply remove the adjectives from the variables.  I'll also use the word "surrender" consistently.

wenyuzhao commented on src/scheduler/scheduler.rs:
```diff
@@ -80,36 +85,61 @@ impl<VM: VMBinding> GCWorkScheduler<VM> {
-        );
-        VM::VMCollection::spawn_gc_thread(tls, GCThreadContext::<VM>::Controller(gc_controller));
+        self.worker_group.create_workers(mmtk);
+        self.worker_group.spawn(tls)
```
`create_workers()` sets the worker state to `Resting`, and `spawn()` immediately updates the state to `Spawned`. Looks like the `Resting` state is not necessary?  Also probably the two-step initialization can be merged.

If `create_workers` also directly spawns the workers, is it possible to get rid of the `Resting` state during this stage? If this is true, we can also rename the Resting state to `Surrendered` to only refer to the state where GC workers are waiting for forking.

wks replied:
Good idea.  I started with creating `GCWorker` instances and spawning them in one step, but I found it was easier to implement by keeping them separate.  But from the perspective of state transition, it is better to make one state transition instead of two in `initialize_collection`.

`Surrendered` is also better than `Resting`, I think.  This will emphasise that `Surrendered` is a special state for forking, rather than a routine state that MMTk needs to go through for all VM bindings (It's only necessary for VMs that need forking).

I'll refactor it to change the two state transitions into one.

wks replied:
Done.

k-sareen commented on src/scheduler/worker_monitor.rs:
```diff
@@ -0,0 +1,238 @@
+    /// Workers wait on this when idle.  Notified if workers have things to do.  That include:
+    /// -   any work packets available, and
+    /// -   any field in `sync.goals.requests` set to true.
+    workers_have_anything_to_do: Condvar,
```
Not sure why you changed the name of the `Condvar`.

wks replied:
The name `work_available` may be ambiguous.  `work` may be interpreted as "work packet", but now the Condvar may also be notified if a "goal" is set.

k-sareen reviewed (commented):
Some major things:

- Why do we need to keep worker state after the GC thread has exited?
- I'm unconvinced that "letting the VM(-binding) handle waiting on threads exiting" is a good API design
- There needs to be at least a few tests using the `MockVM` to ensure this functionality works in the CI
- I think a lot of the comments have gone stale given you've done many iterations of this
- Since this is a significant change, we need to have performance numbers (I guess for OpenJDK) before we can merge this

I've added minor comments here-and-there for spelling etc.

qinsoon commented on src/mmtk.rs:
```diff
@@ -220,6 +220,89 @@ impl<VM: VMBinding> MMTK<VM> {
+    /// This function sends an asynchronous message to GC threads and returns immediately, but it
+    /// is only safe for the VM to call `fork()` after the underlying **native threads** of the GC
+    /// threads have exited.  After calling this function, the VM should wait for their underlying
+    /// native threads to exit in VM-specific manner before calling `fork()`.
```
> For example, I don't register MMTk GC workers into the ART thread list (otherwise they get suspended and all sorts of issues occur) so I would have to maintain an extra list of GC workers which already exists inside mmtk-core, duplicating code/effort.

This sounds like an issue for the binding implementation. How does ART deal with its own GC threads? If they store them in a list, you should put MMTk GC threads to that list as well.

k-sareen replied:
I can go into details in the meeting, but ART is very different. We just cannot register GC threads with the ART threading system. I will essentially have to rewrite the entire threading subsystem (and have rewritten parts of it already).

wks commented on src/scheduler/scheduler.rs:
```diff
@@ -80,36 +85,61 @@ impl<VM: VMBinding> GCWorkScheduler<VM> {
+
+    /// Respawn GC threads after forking.  This will reuse the `GCWorker` instances of stopped
+    /// workers.
+    pub fn respawn_gc_threads_after_forking(self: &Arc<Self>, tls: VMThread) {
```
Yes.  I'll update the doc comment for that.

wks commented on src/memory_manager.rs:
```diff
@@ -485,28 +474,22 @@ pub fn start_control_collector<VM: VMBinding>(
 /// This is a mandatory call for a VM during its boot process once its thread system
-/// is ready. This should only be called once. This call will invoke Collection::spawn_gc_thread()
-/// to create GC threads.
+/// is ready.  This call will invoke `Collection::spawn_gc_thread()`` to create GC threads.
```
Done.

wks commented on tools/tracing/README.md:
```diff
@@ -16,11 +16,13 @@ shipped with the MMTk release you use.
 
--   `mmtk:collection_initialized()`: GC is enabled
+-   `mmtk:collection_initialized()`: All GC worker threads are spawn
+-   `mmtk:collection_uninitialized()`: All GC worker threads exited.  This may happend during
```
Done.  I removed `collection_uninitialized` and added several other trace points.

wks commented on src/scheduler/scheduler.rs:
```diff
@@ -358,29 +388,212 @@ impl<VM: VMBinding> GCWorkScheduler<VM> {
     }
 
+    /// Called when the last worker parked.
+    /// `should_schedule_gc` is true if a mutator requested a GC.
```
That was a stale comment.  I updated it.

wks commented on src/scheduler/scheduler.rs:
```diff
@@ -358,29 +388,212 @@ impl<VM: VMBinding> GCWorkScheduler<VM> {
+        assert!(goals.current().is_none());
+
+        let Some(goal) = goals.poll_next_goal() else {
+            // No reqeusts.  Park this worker, too.
```
Done.

wks commented on src/scheduler/worker_goals.rs:
```diff
@@ -0,0 +1,78 @@
+//! working towards on a high level.
+//!
+//! A "goal" is represented by a `WorkerGoal`.  All workers work towards a single goal at a time.
+//! THe current goal influences the behavior of GC workers, especially the last parked worker.
```
Done.

k-sareen commented:
Ah yes. That makes sense. Dynamic number of threads might not be super important for us to support (but I may be wrong), so we can file an issue and revisit it later at some point.

Re: Implicit vs explicit API. Yeah sure. I understand your stance. I'm fine with the implicit API, but I'll do another pass through the PR later today.

Re: Tests. I think we need some simple CI tests to ensure that this feature does not break due to refactoring(s) in the future.

wks commented:
>     * I'm unconvinced that "letting the VM(-binding) handle waiting on threads exiting" is a good API design

See https://github.com/mmtk/mmtk-core/pull/1067#discussion_r1537251037

> 
>     * There needs to be at least a few tests using the `MockVM` to ensure this functionality works in the CI

I'll add some tests for

-   Triggering GC (`handle_user_collection_request`)
-   Stopping for forking
-   Resuming after forking

>     * I think a lot of the comments have gone stale given you've done many iterations of this

I'll go through interesting places again to see if there are any stale comments.

>     * Since this is a significant change, we need to have performance numbers (I guess for OpenJDK) before we can merge this

We had some performance data here: https://github.com/mmtk/mmtk-core/pull/1062  The performance was improved a little, but very insignificantly.

wks commented:
I have gone through my changes and updated stale comments.  I also added some unit tests.  I think I have addressed all comments except (1) some assertions about `initialize_collection` and allocation (https://github.com/mmtk/mmtk-core/pull/1067#discussion_r1534887007) and (2) creating `GCWorkerShared` later (https://github.com/mmtk/mmtk-core/pull/1067#discussion_r1533219453).  We can do the refactoring after this PR.

I am doing a last-minute performance evaluation on vole.moma to check if there is any regression since the last time.  The result will come out this evening.

k-sareen reviewed (approved):
LGTM. Thanks! The only thing I'm concerned about is how the additional complexity in the scheduler affects @wenyuzhao's work packet/scheduler redesign. I presume PRs for OpenJDK and other bindings to remove the coordinator thread will follow? Also could you fix the typos I highlighted in a previous review?

wks commented:
The result.  On `mole.moma`, lusearch from DaCapo Chopin, 40 invocations, 5 iterations each, 3x min heap size w.r.t. OpenJDK G1, using 5 different plans, comparing the master branch and this PR.

(MMTK_NO_REFERENCE_TYPES is accidentally set to false, and I'll re-run the benchmark without setting it.)

![image](https://github.com/mmtk/mmtk-core/assets/370317/00dbd350-eed9-4843-b90c-a7693678b515)

Plotty: https://squirrel.anu.edu.au/plotty/wks/noproject/#0|vole-2024-04-03-Wed-070951&benchmark^build^invocation^iteration^mmtk_gc&GC^time^time.other^time.stw&|10&iteration^1^4|20&1^invocation|30&1&mmtk_gc&build;build1|40&Histogram%20(with%20CI)^build^mmtk_gc&

The result shows an average 2% reduction in total time across the five different plans, where GenImmix has a 4% reduction and SemiSpace has a 0.5% reduction in total time.  The results in https://github.com/mmtk/mmtk-core/pull/1062 only shows a moderate 0.5% improvement on average.  I'll re-run the benchmark with reference processing disabled to see if it has any impact.

wks commented on src/scheduler/work_bucket.rs:
```diff
@@ -139,6 +139,19 @@ impl<VM: VMBinding> WorkBucket<VM> {
         self.notify_one_worker();
     }
 
+    /// Add a work packet ot this bucket, but do not notify any workers.
```
Done.

wks commented:
Here are the results from two other machines.

lynx.moma (using `taskset -c 0-15` to limit the execution to big cores, only): ([plotty](https://squirrel.anu.edu.au/plotty/wks/noproject/#0|lynx-2024-04-04-Thu-033919&benchmark^build^invocation^iteration^mmtk_gc&GC^time^time.other^time.stw&|10&iteration^1^4|20&1^invocation|30&1&mmtk_gc&build;build1|40&Histogram%20(with%20CI)^build^mmtk_gc&)) 

![image](https://github.com/mmtk/mmtk-core/assets/370317/d19824b4-8f94-49b1-92b3-6a33fb67c7c6)

fisher.moma: ([plotty](https://squirrel.anu.edu.au/plotty/wks/noproject/#0|fisher-2024-04-04-Thu-035645&benchmark^build^invocation^iteration^mmtk_gc&GC^time^time.other^time.stw&|10&iteration^1^4|20&1^invocation|30&1&mmtk_gc&build;build1|41&Histogram%20(with%20CI)^build^mmtk_gc&))

![image](https://github.com/mmtk/mmtk-core/assets/370317/7cdb991f-05bb-41ed-88c0-dbd3969f7ad2)

These results are similar.  The reduction in total time is noticeable for generational plans, but insignificant for non-generational ones.   The total time improvement on lynx and fisher is not as significant as mole.moma.

wks commented:
The extended-tests-openjdk CI test failed because it hung while triggering the first GC when running antlr from DaCapo 2006 using MarkSweep.  However, I can't reproduce this locally.

JikesRVM failed for "Failing instruction starting at f797b0a1 wasn't in RVM address space".

I restarted those tests.

wks commented:
The OpenJDK binding test failed again because it still got stuck while running MarkSweep.  I managed to reproduce it locally.  In `mmtk::policy::marksweepspace::native_ms::block::Block::simple_sweep`, the `cell_size` variable can sometimes be 0, causing the `while` loop to go on forever.

```rs
        let cell_size = self.load_block_cell_size(); // This sometimes returns 0.
        let mut cell = self.start();
        // ...
        while cell + cell_size <= self.start() + Block::BYTES {
            // ...
            cell += cell_size; // If `cell_size` is 0, this loop will go on forever.
        }
```

It is likely an unrelated bug, and I'll report it in a separate issue and try to find the cause later.
